pages/reference/OS/configuration.md

```diff
@@ -0,0 +1,179 @@
+
+__Note:__ The `config.json` file is different from the [`config.txt`](https://www.balena.io/docs/reference/OS/advanced/#configtxt) file also located in the boot partition, which is used by the Raspberry Pi to set device configuration options.
+
+## Using config.json to provision a device
+
+The `config.json` file is located in the root of the `resin-boot` partition, and you may edit this file by mounting a flashed SD card and editing the file directly. Editing this file will only work if you edit before the device's first boot. Once the device has been provisioned, edits to this file have no effect as the contents of the file are copied to the device boot partition at `/mnt/boot/config.json`.
+
+## Sample config.json
+
+The following example provides all available configuration options available in `config.json`. Full details about each value may be found in the [valid fields](#valid-fields) section.
+
+```json
+{
+  "hostname": "my-custom-hostname",
+  "persistentLogging": true,
+  "country": "GB",
+  "ntpServers": "ntp-wwv.nist.gov resinio.pool.ntp.org",
+  "dnsServers": "208.67.222.222 8.8.8.8",
+  "os": {
+    "network": {
+      "connectivity": {
+        "uri": "https://api.balena-cloud.com/connectivity-check",
+        "interval": "300",
+        "response": "optional value in the response"
+      },
+      "wifi": {
+        "randomMacAddressScan": false
+      }
+    },
+    "udevRules": {
+      "56": "ENV{ID_FS_LABEL_ENC}==\"resin-root*\", IMPORT{program}=\"resin_update_state_probe $devnode\", SYMLINK+=\"disk/by-state/$env{RESIN_UPDATE_STATE}\"",
+      "64": "ACTION!=\"add|change\", GOTO=\"modeswitch_rules_end\"\nKERNEL==\"ttyACM*\", ATTRS{idVendor}==\"1546\", ATTRS{idProduct}==\"1146\", TAG+=\"systemd\", ENV{SYSTEMD_WANTS}=\"u-blox-switch@'%E{DEVNAME}'.service\"\nLBEL=\"modeswitch_rules_end\"\n"
+    },
+    "sshKeys": [
+      "ssh-rsa AAAAB3Nza...M2JB balena@macbook-pro",
+      "ssh-rsa AAAAB3Nza...nFTQ balena@zenbook"
+    ]
+  }
+}
+```
+
+## Valid fields
+
+The following fields can be used to modify the behavior of the host OS.
+
+### hostname
+
+(string) The configured hostname of the device, otherwise the device UUID is used.
+
+### persistentLogging
+
+(boolean) Enable or disable persistent logging on the device. Defaults to false.
+
+### country
+
+(string) The country in which the device is operating. Used for setting the WiFi regulatory domain.
+
+### ntpServers
+
+(string) A space-separated list of NTP servers to use for time synchronization. Defaults to `resinio.pool.ntp.org` servers.
+
+### dnsServers
+
+(string) A space-separated list of preferred DNS servers to use for name resolution.
+
+* When `dnsServers` is not defined, or empty, the default value of Google's DNS (8.8.8.8) is used.
+* When `dnsServers` is `"null"` (a string), the system will not alter the dnsServers to the default server as detailed above.
```

Comment:
Does this mean there'd be no DNS resolution at all?

Reply:
No, DNS servers obtained via DHCP (and/or [mentioned in a NetworkManager connection profile](https://www.balena.io/docs/reference/OS/network/2.x/#setting-a-static-ip)?) are also used. Also, I think (hope!) that the following paragraph in the [Network Requirements](https://www.balena.io/docs/reference/OS/network/2.x/#network-requirements) docs is correct, and could be added / copied (?) to this section (or have it only here, and change the network requirements to link here...)

> Google's Public DNS server at 8.8.8.8 is used by default and in addition to DNS servers obtained via DHCP (balenaOS may issue queries to multiple DNS servers simultaneously, for the quickest response to be used). If additional DNS servers are configured via DHCP or other means, it is OK for the local network to block requests to 8.8.8.8. To avoid any requests being made to 8.8.8.8 by balenaOS, modify the dnsServers setting in config.json.

So maybe actually the three bullet points would be more informative if rewritten as follows, though this needs testing to confirm, or source code to be read, or expert advice from perhaps @majorz or @alexgg (Zubair is away for a while...)
* When `dnsServers` is not defined, or empty, Google's DNS server (8.8.8.8) is added to the list of DNS servers obtained via DHCP or statically configured in a NetworkManager connection profile.
* When `dnsServers` is `"null"` (a string), Google's DNS server (8.8.8.8) will NOT be added as described above.
* When `dnsServers` is defined and not `"null"`, the listed servers will _replace_ any servers obtained via DCHP or statically configured via a NetworkManager connection profile.

Reply:
Hi @pdcastro, how I understand the DNS resolution in Balena is that it uses dnsmasq as a local cache and forwarder.
/etc/resolv.conf is symlinked to resolv.dnsmasq
and dnsmasq is started with:
`-r /etc/resolv.dnsmasq -z --servers-file=/run/dnsmasq.servers`
`/etc/resolv.dnsmasq` is a symlink to `/var/run/resolvconf/interface/NetworkManager`, which is the file that NetworkManager uses to store the DHCP or statically configured DNS server. This is the first place where dnsmasq lookes for upstream servers.
And `/run/dnsmasq.servers` is the file that is configured from config.json with either the user provided servers or 8.8.8.8.
So dnsServers only affects the contents of `/run/dnsmasq.servers`, and the NM DHCP or static configuration will always be there.
So I would rephase the above to:

- When `dnsServers` is not defined, or empty, Google's DNS server (8.8.8.8) is added to the list of DNS servers obtained via DHCP or statically configured in a NetworkManager connection profile.
- When `dnsServers` is "null" (a string), Google's DNS server (8.8.8.8) will NOT be added as described above.
- When `dnsServers` is defined and not "null", the listed servers will be added to the list of servers obtained via DCHP or statically configured via a NetworkManager connection profile.

Reply:
Thank you for digging this @alexgg! 👍 We are finally getting more clarity on these points. It surprises me that `--all-servers` isn't the default behaviour, as I believe this is the "support knowledge" that we share with users. @majorz, are you similarly surprised? :-) 
In this case, we need to update the Network Requirements page as well, as it is incorrect.

Reply:
There is some additional material here on this issue just published via @hedss WIP masterclass https://hackmd.io/7z1NBeDoTy2rqchIC9ogHg?view#52-DNS-Issues

Reply:
Opened an issue to track this https://github.com/balena-io/docs/issues/1283